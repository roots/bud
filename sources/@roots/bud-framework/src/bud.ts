import {bind} from '@roots/bud-support/decorators'
import isFunction from '@roots/bud-support/lodash/isFunction'
import isNull from '@roots/bud-support/lodash/isNull'
import isString from '@roots/bud-support/lodash/isString'
import isUndefined from '@roots/bud-support/lodash/isUndefined'

import {bootstrap, LIFECYCLE_EVENT_MAP} from './lifecycle/bootstrap.js'
import type * as methods from './methods/index.js'
import type {Module} from './module.js'
import type {Service} from './service.js'
import type ConsoleBuffer from './services/console.js'
import type FS from './services/fs.js'
import type * as Options from './types/options/index.js'
import type * as Registry from './types/registry/index.js'
import type Hooks from './types/services/hooks/index.js'
import type * as Services from './types/services/index.js'
import Value from './value.js'

/**
 * Framework abstract
 */
export class Bud {
  /**
   * Context
   *
   * @public
   */
  public context:
    | Options.CommandContext
    | Options.CLIContext
    | Options.Context

  public isCLI(): this is Bud & {context: Options.CommandContext} {
    return `args` in this.context
  }

  /**
   * Implementation
   *
   * @public
   */
  public implementation: new () => Bud

  /**
   * Compilation mode
   *
   * @remarks
   * Either `production` or `development`.
   *
   * @readonly
   * @defaultValue `production`
   * @public
   */
  public get mode(): `development` | `production` {
    return this.context.mode ?? `production`
  }

  /**
   * Name
   *
   * @readonly
   * @public
   */
  public get label() {
    return this.context.label
  }

  /**
   * Parent {@link Bud} instance
   *
   * @readonly
   * @public
   */
  public get root() {
    return this.context.root ?? this
  }

  /**
   * True when {@link Bud.mode} is `production`
   *
   * @public
   */
  public get isProduction(): boolean {
    return this.mode == `production`
  }

  /**
   * True when {@link Bud.mode} is `development`
   *
   * @public
   */
  public get isDevelopment(): boolean {
    return this.mode == `development`
  }

  /**
   * True when current instance is the parent instance
   *
   * @readonly
   * @public
   */
  public get isRoot(): boolean {
    return this.root.label === this.label
  }

  /**
   * True when current instance is a child instance
   *
   * @readonly
   * @public
   */
  public get isChild(): boolean {
    return this.root.label !== this.label
  }

  /**
   * {@link Bud} instances
   *
   * @public
   */
  public children?: Record<string, Bud>

  /**
   * True when child compilers
   *
   * @readonly
   * @public
   */
  public get hasChildren(): boolean {
    return (
      !isUndefined(this.children) &&
      !isNull(this.children) &&
      Object.entries(this.children).length > 0
    )
  }

  public consoleBuffer: ConsoleBuffer

  public fs: FS

  public module: Module

  public services: Array<string> = []

  public api: Services.Api

  public build: Services.Build.Service

  public cache: Services.Cache.Service

  public compiler: Services.Compiler.Service

  public dashboard: Services.Dashboard.Service

  public env: Services.Env

  public extensions: Services.Extensions.Service

  public hooks: Hooks

  public project: Services.Project.Service

  public server: Services.Server.Service

  public declare after: methods.after

  public declare maybeCall: methods.maybeCall

  public declare close: methods.close

  public declare container: methods.container

  public declare get: methods.get

  public declare glob: methods.glob

  public declare globSync: methods.globSync

  public declare path: methods.path

  public declare pipe: methods.pipe

  public declare processConfigs: methods.processConfigs

  public declare publicPath: methods.publicPath

  public declare relPath: methods.relPath

  public declare run: methods.run

  public declare setPath: methods.setPath

  public declare setPublicPath: methods.setPublicPath

  public declare sequence: methods.sequence

  public declare sequenceSync: methods.sequenceSync

  public declare sh: methods.sh

  public declare tap: methods.tap

  public declare tapAsync: methods.tapAsync

  public declare when: methods.when

  public declare bindMethod: methods.bindMethod

  /**
   * @deprecated - use {@link bud.fs.json | bud.fs.json}
   */
  public json: FS['json']

  /**
   * @deprecated - use {@link bud.fs.yml | bud.fs.yml}
   */
  public yml: FS['yml']

  /**
   * Value helper
   *
   * @public
   */
  public value = Value

  /**
   * Creates a child with `bud.create` but returns the parent instance
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async make(
    request: Partial<Options.Context> | string,
    tap?: (app: Bud) => Promise<unknown>,
  ) {
    if (!this.isRoot)
      return this.fatal(
        new Error(
          `Child instances should be produced from the root context`,
        ),
      )

    const context: Options.Context = isString(request)
      ? {...this.context, label: request, root: this}
      : {...this.context, ...request, root: this}

    if (
      this.isCLI() &&
      !isUndefined(this.context.args.filter) &&
      !this.context.args.filter.includes(context.label)
    ) {
      this.log(
        `skipping child instance based on --filter flag:`,
        context.label,
      )
      return this
    }

    if (this.children && this.children[context.label]) {
      this.log(`returning requested child instance:`, context.label)
      return this.get(context.label)
    }

    this.log(`instantiating new bud instance`)
    const child = await new Bud().lifecycle(context)

    if (!this.children) this.children = {[context.label]: child}
    else this.children[context.label] = child

    this.get(context.label).hooks.on(
      `build.dependencies`,
      typeof request !== `string` && request.dependsOn
        ? request.dependsOn
        : Object.values(this.children)
            .map(({label}) => label)
            .filter(label => label !== context.label),
    )

    if (tap) await tap(this.get(context.label))
    await this.get(context.label)?.api.processQueue()

    return this
  }

  @bind
  public async lifecycle(context: Options.Context): Promise<Bud> {
    await bootstrap.bind(this)({...context})

    Object.entries(LIFECYCLE_EVENT_MAP).map(
      ([eventHandle, callbackName]: [
        keyof Registry.EventsStore,
        keyof Service,
      ]) =>
        this.services
          .map(service => [service, this[service]])
          .map(([label, service]) => {
            if (!service) {
              this.error(`service not found: ${label}`, this.services)
            }

            if (!isFunction(service[callbackName])) return

            this.hooks.action(
              eventHandle,
              service[callbackName].bind(service),
            )
            this.success(
              `registered service callback:`,
              `${label}.${callbackName}`,
            )
          }),
    )

    await [
      `init`,
      `bootstrap`,
      `bootstrapped`,
      `register`,
      `registered`,
      `boot`,
      `booted`,
    ].reduce(async (promised, event: keyof Registry.EventsStore) => {
      await promised
      await this.hooks.fire(event)
    }, Promise.resolve())

    return this
  }

  /**
   * Log a message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public log(...messages: any[]) {
    this.context.logger.log(...messages)

    return this
  }

  /**
   * Log an `info` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public info(...messages: any[]) {
    this.context.logger.info(...messages)

    return this
  }

  /**
   * Log a `success` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public success(...messages: any[]) {
    this.context.logger.success(...messages)
    return this
  }

  /**
   * Log a `warning` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public warn(...messages: any[]) {
    this.context.logger.warn(...messages)

    return this
  }

  /**
   * Log an error.
   *
   * @remarks
   * In `production` this error is treated as fatal
   * and will kill the process.
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public error(...messages: Array<any>): Bud {
    this.context.logger.error(...messages)
    return this
  }

  /**
   * Log and display an error.
   *
   * @remarks
   * This will always kill the process
   *
   * @public
   * @decorator `@bind`
   * @throws fatal error
   */
  @bind
  public fatal(error: Error) {
    this.context.logger.fatal(error)
    return this
  }
}
