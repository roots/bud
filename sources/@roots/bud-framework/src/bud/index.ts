import type {
  Api,
  Build,
  Cache,
  Compiler,
  Context,
  Dashboard,
  Env,
  Extensions,
  Hooks,
  Project,
  Server,
  Service,
} from '@roots/bud-framework'

import methods from '@roots/bud-framework/methods'
import {bind} from '@roots/bud-support/decorators/bind'
import {BudError, InputError} from '@roots/bud-support/errors'
import isFunction from '@roots/bud-support/isFunction'
import isString from '@roots/bud-support/isString'
import isUndefined from '@roots/bud-support/isUndefined'
import logger from '@roots/bud-support/logger'

import type {FS} from '../fs.js'
import type {Module} from '../module.js'
import type {Notifier} from '../notifier.js'
import type {EventsStore} from '../registry/index.js'

import {bootstrap} from '../bootstrap.js'

/**
 * Bud core class
 */
export class Bud {
  /**
   * Promised tasks
   */
  public promised: Array<(bud: Bud) => Promise<any>> = []

  public declare addConfig: typeof methods.addConfig

  public declare after: typeof methods.after

  public declare api: Service & Api

  public declare bindFacade: typeof methods.bindFacade

  public declare build: Service & Build

  public declare cache: Service & Cache

  /**
   * {@link Bud} instances
   */
  public declare children: Record<string, Bud>

  public declare close: typeof methods.close

  public declare compiler?: Service & Compiler

  public declare container: typeof methods.container

  public declare context: Context

  public declare dashboard: Service & Dashboard

  public declare env: Service & Env

  public declare extensions: Service & Extensions

  public declare fs: FS

  public declare get: typeof methods.get

  public declare glob: typeof methods.glob

  public declare globSync: typeof methods.globSync

  public declare hooks: Service & Hooks

  /**
   * {@link Bud} Implementation
   * @internal
   */
  public declare implementation: new () => Bud

  /**
   * @deprecated Use {@link bud.fs}
   * @readonly
   */
  public declare json: FS['json']

  public declare maybeCall: typeof methods.maybeCall

  public declare module: Module

  public declare notifier: Notifier

  public declare path: typeof methods.path

  public declare pipe: typeof methods.pipe

  public declare processConfigs: typeof methods.processConfigs

  public declare project: Project

  public declare publicPath: typeof methods.publicPath

  public declare relPath: typeof methods.relPath

  public declare run: typeof methods.run

  public declare sequence: typeof methods.sequence

  public declare sequenceSync: typeof methods.sequenceSync

  public declare server?: Service & Server

  public declare setPath: typeof methods.setPath

  public declare setPublicPath: typeof methods.setPublicPath

  public declare sh: typeof methods.sh

  public declare tap: typeof methods.tap

  public declare tapAsync: typeof methods.tapAsync

  public declare when: typeof methods.when

  /**
   * @deprecated Use {@link Bud.fs}
   * @readonly
   */
  public declare yml: FS['yml']

  /**
   * True when child compilers
   * @readonly
   */
  public get hasChildren(): boolean {
    return this.children && Object.entries(this.children).length > 0
  }

  /**
   * True when current instance is a child instance
   * @readonly
   */
  public get isChild(): boolean {
    return this.root?.context?.label !== this.context?.label
  }

  /**
   * True when {@link Bud.mode} is `development`
   * @readonly
   */
  public get isDevelopment(): boolean {
    return this.mode === `development`
  }

  /**
   * True when {@link Bud.mode} is `production`
   * @readonly
   */
  public get isProduction(): boolean {
    return this.mode === `production`
  }

  /**
   * True when current instance is the parent instance
   * @readonly
   */
  public get isRoot(): boolean {
    return this.root?.context?.label === this.context?.label
  }

  /**
   * Label
   * @readonly
   */
  public get label() {
    return this.context?.label ?? `bud`
  }

  /**
   * Compilation mode
   *
   * @remarks
   * Either `production` or `development`.
   *
   * @readonly
   * @defaultValue `production`
   */
  public get mode(): `development` | `production` {
    return this.context?.mode ?? `production`
  }

  /**
   * Constructor
   */
  public constructor(context?: Context) {
    if (context) this.context = {...context}

    this.set(`implementation`, this.constructor as any)

    Object.entries(methods).map(([k, v]) =>
      this.set(k as any, v.bind(this)),
    )
  }

  /**
   * Boot application services
   */
  @bind
  public async boot() {
    await this.executeServiceCallbacks(`boot`).catch(this.catch)
  }

  /**
   * Bootstrap the application
   */
  @bind
  public async bootstrap() {
    await this.executeServiceCallbacks(`bootstrap`).catch(this.catch)
  }

  /**
   * Error handler
   */
  @bind
  public catch(error: Error): never {
    const normalizedError = BudError.normalize(error)
    if (!normalizedError.instance && this?.isChild)
      normalizedError.instance = this.label
    throw normalizedError
  }

  /**
   * Execute service callbacks for a given stage
   * @internal
   */
  @bind
  public async executeServiceCallbacks(
    stage: `${keyof EventsStore & string}`,
  ): Promise<void> {
    await this.resolvePromises().catch(this.catch)
    await this.hooks.fire(stage, this)
  }

  /**
   * Bud initialize
   */
  @bind
  public async initialize(context?: Context): Promise<Bud> {
    if (context) this.context = {...(this.context ?? {}), ...context}

    await bootstrap(this).catch(this.catch)

    await this.bootstrap().catch(this.catch)
    await this.register().catch(this.catch)
    await this.boot().catch(this.catch)
    await this.executeServiceCallbacks(`config.before`).catch(this.catch)

    return this
  }

  /**
   * Creates a child and returns the parent instance
   */
  @bind
  public async make(
    request: Partial<Context> | string,
    setupFn?: (app: Bud) => Promise<unknown>,
  ) {
    if (!this.isRoot) {
      return this.catch(
        InputError.normalize(
          `bud.make: must be called from the root context`,
        ),
      )
    }

    const context: Context = isString(request)
      ? {...this.context, label: request, root: this}
      : {...this.context, ...request, root: this}

    if (isUndefined(context.label)) {
      return this.catch(
        InputError.normalize(`bud.make: context.label must be a string`),
      )
    }

    if (
      !isUndefined(this.context.filter) &&
      !this.context.filter.includes(context.label)
    ) {
      logger
        .scope(`make`)
        .log(
          `skipping child instance based on --filter flag:`,
          context.label,
        )

      return this
    }

    if (!this.children) this.children = {}

    if (this.children[context.label]) {
      return this.catch(
        InputError.normalize(
          `bud.make: child instance ${context.label} already exists`,
        ),
      )
    }

    logger.scope(`make`).log(`instantiating ${context.label}`)

    this.children[context.label] =
      await new this.implementation().initialize({
        ...context,
      })
    if (setupFn) await setupFn(this.children[context.label])

    await this.children[context.label].resolvePromises().catch(this.catch)

    if (typeof request !== `string` && request.dependsOn) {
      this.get(context.label)?.hooks.on(
        `build.dependencies`,
        request.dependsOn,
      )
    }

    return this
  }

  /**
   * Await all promised tasks
   */
  @bind
  public promise(promise: (bud: Bud) => Promise<unknown>): Bud {
    this.promised.push(promise)
    return this
  }

  @bind
  public async resolvePromises() {
    if (this.promised.length === 0) return

    const promises = this.promised

    this.promised = []

    await promises.reduce(async (promised, promise) => {
      await promised
      await promise(this)
    }, Promise.resolve())
  }

  /**
   * Register application services
   */
  @bind
  public async register() {
    await this.executeServiceCallbacks(`register`).catch(this.catch)
  }

  /**
   * Parent {@link Bud} instance
   * @readonly
   */
  public get root(): Bud {
    return this.context?.root ?? this
  }

  /**
   * Set a value on the current instance
   * @param key - key
   * @param value - value
   * @param bind - bind value to current instance (default: true, if bindable)
   */
  @bind
  public set<K extends `${keyof Bud & string}`>(
    key: K,
    input: Bud[K],
    bind: boolean = true,
  ): Bud {
    const toBind = bind && isFunction(input) && `bind` in input

    const value = toBind ? input.bind(this) : input

    Object.defineProperty(this, key, {
      configurable: true,
      enumerable: true,
      value,
      writable: true,
    })

    return this
  }

  /**
   * Log message
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public log(...messages: Array<any>) {
    logger.scope(this.label).log(...messages)
    return this
  }

  /**
   * Log success
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public success(...messages: Array<any>) {
    logger.scope(this.label).log(...messages)
    return this
  }
  /**
   * Log warning
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public warn(...messages: Array<any>) {
    logger.scope(this.label).warn(...messages)
    return this
  }

  /**
   * Log info
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public info(...messages: Array<any>) {
    logger.scope(this.label).info(...messages)
    return this
  }

  /**
   * Log error
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public error(...messages: Array<any>) {
    logger.scope(this.label).error(...messages)
    return this
  }
}
