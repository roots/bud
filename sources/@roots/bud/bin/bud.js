#!/usr/bin/env node
// @ts-check

const {Cli, Builtins} = require('clipanion');
const {BuildCommand, CleanCommand, DevCommand, DoctorCommand, InstallCommand} = require('../lib/cjs/cli/index.js');
const { fs } = require('@roots/bud-support')

const args = process.argv.splice(2);

(async () => {
  const manifest = await fs.readJson(`${__dirname}/../package.json`)
  const name = manifest.name.split('/').pop()

  const cli = new Cli({
    binaryLabel: name,
    binaryName: name,
    binaryVersion:manifest.version,
    enableColors: true,
  })

  cli.register(Builtins.HelpCommand)
  cli.register(Builtins.DefinitionsCommand)
  cli.register(Builtins.VersionCommand)
  cli.register(BuildCommand);
  cli.register(CleanCommand);
  cli.register(DevCommand);
  cli.register(DoctorCommand);
  cli.register(InstallCommand)

  cli.run(args);
})();
