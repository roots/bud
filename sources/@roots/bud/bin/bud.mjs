#!/usr/bin/env node

const run = async () => {
  await import(`../lib/cli/index.js`);
};

run();
