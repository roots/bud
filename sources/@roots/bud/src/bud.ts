import * as Framework from '@roots/bud-framework'

/**
 * ⚡️ Bud
 *
 * @public
 */
export default class Bud extends Framework.Bud {
  public constructor() {
    super(Bud)
    this.implementation = Bud
  }
}
