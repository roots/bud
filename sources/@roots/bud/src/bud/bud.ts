import type Api from '@roots/bud-api'
import type Build from '@roots/bud-build'
import type Cache from '@roots/bud-cache'
import type Compiler from '@roots/bud-compiler'
import type Dashboard from '@roots/bud-dashboard'
import type Extensions from '@roots/bud-extensions'
import {Bud as Framework} from '@roots/bud-framework'
import {Service, ServiceContainer} from '@roots/bud-framework'
import type Hooks from '@roots/bud-hooks'
import type Server from '@roots/bud-server'

/**
 * ## ⚡️ Bud
 *
 * {@link https://bud.js.org/ bud.js}
 */
export class Bud extends Framework {
  public override implementation = Bud

  public declare api: Api

  public declare build: Build

  public declare cache: Cache

  public declare compiler: Compiler

  public declare dashboard: Dashboard

  public declare extensions: Extensions

  public declare hooks: Hooks

  public declare server: Server
}

export {Service, ServiceContainer}
