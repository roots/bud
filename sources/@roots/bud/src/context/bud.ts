import {bind} from '@roots/bud-support/decorators'
import * as fs from '@roots/bud-support/fs'
import {dirname, join, resolve, sep} from 'node:path/posix'
import {fileURLToPath} from 'node:url'

/**
 * Application context
 *
 * @public
 */
export default class Bud {
  public data: {
    label: string
    basedir: string
    version: string
    manifestPath: string
  } = {
    label: null,
    basedir: null,
    version: null,
    manifestPath: null,
  }

  /**
   * Find application manifest
   *
   * @public
   */
  @bind
  public async find(): Promise<this> {
    const resolvedPath = dirname(fileURLToPath(import.meta.url))
    this.data.manifestPath = resolve(
      join(resolvedPath, `..`, `..`, `package.json`),
    )
    this.data.basedir = dirname(resolve(this.data.manifestPath))
    await this.handleFindResults(this.data.manifestPath)

    return this
  }

  /**
   * Handle file results
   *
   * @public
   */
  @bind
  public async handleFindResults(path: string): Promise<this> {
    const manifest = await fs.readJson(path)

    this.data.label = manifest.name.split(sep).pop()
    this.data.version = manifest.version

    return this
  }
}
