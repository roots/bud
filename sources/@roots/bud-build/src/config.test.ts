import {Bud, factory} from '@repo/test-kit/bud'

import Build from './index.js'

describe(`bud.build.config`, function () {
  let bud: Bud
  let build: Build

  beforeAll(async () => {
    bud = await factory()
    build = new Build(
      // @ts-ignore
      bud,
    )

    await build.register()
    await build.make()
  })

  it(`should not include deprecated properties`, () => {
    expect(build.config.hasOwnProperty(`devServer`)).toBe(false)
    expect(build.config.hasOwnProperty(`unsafeCache`)).toBe(false)
  })

  it(`should have expected bail default`, () => {
    expect(build.config.bail).toEqual(true)
  })

  it(`should have expected cache default`, () => {
    const {cache}: any = build.config

    expect(cache.type).toStrictEqual(`filesystem`)

    expect(cache.buildDependencies.config).toEqual(
      expect.arrayContaining([
        expect.stringContaining(`project/.eslintrc.js`),
        expect.stringContaining(`project/bud.config.mjs`),
        expect.stringContaining(`project/docker-compose.yml`),
        expect.stringContaining(`project/package.json`),
        expect.stringContaining(`project/tailwind.config.js`),
        expect.stringContaining(`project/tsconfig.json`),
        expect.stringContaining(`project/webpack.config.mjs`),
      ]),
    )

    expect(cache.cacheDirectory).toStrictEqual(
      expect.stringContaining(`.budfiles`),
    )

    expect(cache.version).toStrictEqual(expect.any(String))
  })

  it(`should have expected context default`, () => {
    expect(build.config.context).toEqual(bud.path())
  })

  it(`should have expected devtool default`, () => {
    expect(build.config.devtool).toBe(false)
  })

  it(`should have expected entry default`, () => {
    expect(build.config.entry).toBeUndefined()
  })

  it(`should have expected mode default`, () => {
    expect(build.config.mode).toEqual(`production`)
  })

  it(`should have expected name default`, () => {
    expect(build.config.name).toEqual(`@tests/project`)
  })

  it(`should have expected node default`, () => {
    expect(build.config.node).toEqual(false)
  })

  it(`should have expected optimization.minimize default`, () => {
    expect(build.config.optimization?.minimize).toEqual(false)
  })

  it(`should have expected optimization.emitOnErrors default`, () => {
    expect((build.config.optimization as any).emitOnErrors).toEqual(false)
  })

  it(`should have expected optimization.runtimeChunk default`, () => {
    expect(build.config.optimization?.runtimeChunk).toBeUndefined()
  })

  it(`should have expected profile default`, () => {
    expect(build.config.profile).toBeUndefined()
  })

  it(`should have expected resolve.alias default`, () => {
    expect(build.config.resolve?.alias).toEqual({
      '@dist': bud.path(`@dist`),
      '@src': bud.path(`@src`),
    })
  })

  it(`should have expected resolve.extensions default`, () => {
    expect(build.config.resolve?.extensions).toMatchSnapshot([
      `.mjs`,
      `.js`,
      `.jsx`,
      `.css`,
      `.json`,
      `.wasm`,
      `.yml`,
    ])
  })

  it(`should have expected target default`, () => {
    expect(build.config.target).toMatch(/browserslist.*/)
  })

  it(`should have expected watch default`, () => {
    expect(build.config.watch).toBeUndefined()
  })

  it(`should have expected watchOptions default`, () => {
    expect(build.config.watchOptions).toBeUndefined()
  })

  it(`should have expected plugins`, () => {
    const plugins = build.config.plugins?.map(
      plugin => plugin.constructor.name,
    )
    expect(plugins).toContain(`EntrypointsWebpackPlugin`)
    expect(plugins).toContain(`WebpackManifestPlugin`)
    expect(plugins).toContain(`ESLintWebpackPlugin`)
    expect(plugins).toContain(`DefinePlugin`)
    expect(plugins).toContain(`MiniCssExtractPlugin`)
  })

  it(`should have expected default requireEnsure rule`, () => {
    if (!build.config.module?.rules?.length) throw new Error()
    expect(build.config.module.rules[0]).toEqual(
      expect.objectContaining({
        parser: {
          requireEnsure: false,
        },
      }),
    )
  })
})
