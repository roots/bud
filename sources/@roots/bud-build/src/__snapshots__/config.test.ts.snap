// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bud.build.config should have expected resolve.extensions default 1`] = `
[
  ".mjs",
  ".js",
  ".jsx",
  ".css",
  ".json",
  ".wasm",
  ".yml",
]
`;
